src/libsyntax/test.rs

```diff
@@ -690,19 +733,65 @@ fn mk_test_desc_and_fn_rec(cx: &TestCtxt, test: &Test) -> P<ast::Expr> {
              field("should_panic", fail_expr),
              field("allow_fail", allow_fail_expr)]);
 
-
-    let mut visible_path = match cx.toplevel_reexport {
-        Some(id) => vec![id],
+    let mut visible_path = vec![];
+    if cx.features.extern_absolute_paths {
+        visible_path.push(keywords::Crate.ident());
+    }
+    match cx.toplevel_reexport {
+        Some(id) => visible_path.push(id),
         None => {
             let diag = cx.span_diagnostic;
             diag.bug("expected to find top-level re-export name, but found None");
         }
     };
     visible_path.extend(path);
 
-    let fn_expr = ecx.expr_path(ecx.path_global(span, visible_path));
+    // If termination feature is enabled, create a wrapper that invokes the fn
+    // like this:
+    //
+    //     fn wrapper() {
+    //         assert_eq!(0, real_function().report());
+    //     }
+    //
+    // and then put a reference to `wrapper` into the test descriptor. Otherwise,
+    // just put a direct reference to `real_function`.
+    let fn_expr = {
+        let base_fn_expr = ecx.expr_path(ecx.path_global(span, visible_path));
+        if cx.features.termination_trait {
+            // ::std::Termination::assert_unit_test_successful
+            let assert_unit_test_successful = ecx.path_global(
+                span,
+                vec![
+                    ecx.ident_of("std"),
+                    ecx.ident_of("Termination"),
+                    ecx.ident_of("assert_unit_test_successful"),
+                ],
+            );
+            // || {..}
+            ecx.lambda(
+                span,
+                vec![],
+                // ::std::Termination::assert_unit_test_successful(..)
```

Comment:
Can't this function be in `libtest`? I do think I like the closure though - it gets coerced into a `fn` pointer, right?

Reply:
Can you say more about what you think I should do here? You want me to generate a closure that invokes a fn in `libtest` that invokes `std::Termination::assert_unit_test_successful`? (Or I guess potentially remove `assert_unit_test_successful`? I'd rather not do the latter, but then it is an ad-hoc extension to the spec. It's one that makes a lot of sense to me though, since I imagine that people may want to write custom result types that print out differently in unit test failures than otherwise. But I guess there is no strong reason that said people shouldn't use a different type in their unit tests.)

Reply:
I think there should be a trait in `libtest` that can be specialized if need be in the future.

Reply:
ok, so roughly what there is no, but indirecting through some random trait instead, and that trait is implemented for all `T: Termination`?

Reply:
That sounds like a good default, yes. I'd really want to keep testing infrastructure out of `libstd` where possible.

Reply:
@eddyb I guess it'd be mildly easier (and equally future proof) to add a top-level function to libtest, like `assert_unit_test_success(foo())`. Internally, it can use `Termination` for now -- we can just use it indiscriminately actually, since `Termination` is defined for `()`.

Reply:
But ok, I'll add something like that.